[WIP] src,tools: output /ayo executable, related fork impl

Fishrock123 opened the pull request:
This allows you to build an `ayo` executable. Tests pass.

No windows work yet. I'm sure there are bits I forgot.

Fishrock123 commented:
> did iojs have a `node` alias?

Yes, I'd been thinking about that and I think the `out/Release/*` stuff is incorrect, we should use `node` there and just create a final symlink. I'll dig up iojs Makefile code for that when I have time unless someone else wants to.

ghost commented:
@Fishrock123 sounds good. do you want to do the changes in this PR, or should we merge this and then create a new PR some other time?

Fishrock123 commented:
Please avoid merging my PRs for me in most cases - often the content is not final until I'm ready. 😅 

In this case, merging this now will only create __more__ conflicts, which is going to be a issue and there are some areas in which we should take care to make ones that wouldn't be too difficult to resolve in the future (hopefully).

varjmes commented:
I put a [WIP] on this, so no-one thinks they should merge it.

refack commented:
@Fishrock123 would you mind if I added changes to `vcbuild.bat`?

Fishrock123 commented:
@refack That'd be excellent. I'd prefer the final symlink path - I'll be changing the rest over to that tonight unless someone jumps on it before me.

refack commented:
> @refack That'd be excellent. I'd prefer the final symlink path - I'll be changing the rest over to that tonight unless someone jumps on it before me.

Just so I'm clear. Build as `ayo`, then add a symlink named `node` that points to `ayo`, or the other way around?

Fishrock123 commented:
I have pushed up what I think will be a bit better for now - it doesn't yet produce a `node` symlink though.

Fishrock123 commented:
@refack have you made any progress?

I'm going to add two more commits here, particularly, adapting https://github.com/nodejs/node/commit/3f8b76ebd5599737ba99b00cfd54b5751c40b512 and https://github.com/nodejs/node/commit/72f1b348b021269f98783aee2f0e89500a20231c

Fishrock123 committed:
tools: output /ayo executable

PR-URL: https://github.com/ayojs/ayo/pull/10
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: olivia <olivia@fastmail.com>

Fishrock123 committed:
src: rename to ayo

PR-URL: https://github.com/ayojs/ayo/pull/10
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: olivia <olivia@fastmail.com>

Fishrock123 committed:
doc: rename node.1 -> ayo.1

We can update the contents later.

Refs: 3f8b76ebd5599737ba99b00cfd54b5751c40b512

PR-URL: https://github.com/ayojs/ayo/pull/10
Reviewed-By: Anna Henningsen <anna@addaleax.net>

Fishrock123 committed:
install: install ayo -> node compat symlink

Install a symlink for compatibility with scripts that expect a binary
called `node` or have a /path/to/node shebang.  Only done on UNIX
platforms because symlinks on Windows require elevated privileges.

Refs: 72f1b348b021269f98783aee2f0e89500a20231c

PR-URL: https://github.com/ayojs/ayo/pull/10
Reviewed-By: Anna Henningsen <anna@addaleax.net>

Fishrock123 commented:
@addaleax @pup ptal

@refack maybe we can do that in another PR

addaleax commented on Makefile:
```diff
@@ -40,9 +40,12 @@ BUILDTYPE_LOWER := $(shell echo $(BUILDTYPE) | tr '[A-Z]' '[a-z]')
 
-NODE_EXE = node$(EXEEXT)
+# Build to /node for easy cross-repo compat.
+BUILD_DIR_NAME = node
```
I have to admit I don’t quite get why this is called `DIR_NAME` and why `$(EXEEXT)` is removed? I mean, it’s still the executable name in the build dir, right…

Fishrock123 commented:
@addaleax good catch, should be addressed now

addaleax commented:
Thanks! I’d suggest merging once CI is back

(I am **so** happy we have Travis here)

refack commented:
I'll follow up with a new PR for `vcbuild.bat`
